lightning/src/ln/onchain_utils.rs

```diff
@@ -1,8 +1,741 @@
+		HolderFundingOutput {
+			funding_redeemscript,
+		}
+	}
+}
+
+impl_writeable_len_match!(HolderFundingOutput, {
+	{ HolderFundingOutput { ref funding_redeemscript }, 2 + funding_redeemscript.len() }
+	}, {
+	funding_redeemscript
+});
+
+/// A wrapper encapsulating all in-protocol differing outputs categories.
+///
+/// The generic API offers access to outputs attributes (amount, weight, category) or allow
+/// transformation such as finalizing an input claiming the output.
+#[derive(Clone, PartialEq)]
+pub(crate) enum PackageSolvingData {
+	RevokedOutput(RevokedOutput),
+	RevokedHTLCOutput(RevokedHTLCOutput),
+	CounterpartyOfferedHTLCOutput(CounterpartyOfferedHTLCOutput),
+	CounterpartyReceivedHTLCOutput(CounterpartyReceivedHTLCOutput),
+	HolderHTLCOutput(HolderHTLCOutput),
+	HolderFundingOutput(HolderFundingOutput),
+}
+
+impl PackageSolvingData {
+	fn amount(&self) -> u64 {
+		let amt = match self {
+			PackageSolvingData::RevokedOutput(ref outp) => { outp.amount },
+			PackageSolvingData::RevokedHTLCOutput(ref outp) => { outp.amount },
+			PackageSolvingData::CounterpartyOfferedHTLCOutput(ref outp) => { outp.htlc.amount_msat / 1000 },
+			PackageSolvingData::CounterpartyReceivedHTLCOutput(ref outp) => { outp.htlc.amount_msat / 1000 },
+			// Note: Currently, amounts of holder outputs spending witnesses aren't used
+			// as we can't malleate spending package to increase their feerate. This
+			// should change with the remaining anchor output patchset.
+			PackageSolvingData::HolderHTLCOutput(..) => { 0 },
+			PackageSolvingData::HolderFundingOutput(..) => { 0 },
+		};
+		amt
+	}
+	fn weight(&self) -> usize {
+		let weight = match self {
+			PackageSolvingData::RevokedOutput(ref outp) => { outp.weight as usize },
+			PackageSolvingData::RevokedHTLCOutput(ref outp) => { outp.weight as usize },
+			PackageSolvingData::CounterpartyOfferedHTLCOutput(..) => { WEIGHT_OFFERED_HTLC as usize },
+			PackageSolvingData::CounterpartyReceivedHTLCOutput(..) => { WEIGHT_RECEIVED_HTLC as usize },
+			// Note: Currently, weights of holder outputs spending witnesses aren't used
+			// as we can't malleate spending package to increase their feerate. This
+			// should change with the remaining anchor output patchset.
+			PackageSolvingData::HolderHTLCOutput(..) => { debug_assert!(false); 0 },
+			PackageSolvingData::HolderFundingOutput(..) => { debug_assert!(false); 0 },
+		};
+		weight
+	}
+	fn category(&self) -> u8 {
+		let category = match self {
+			// Note, Revoked*Output are sharing the same category as their aggregable
+			// doesn't raise issue, assuming timelocks are respected.
+			PackageSolvingData::RevokedOutput(..) => { 0 },
+			PackageSolvingData::RevokedHTLCOutput(..) => { 0 },
+			PackageSolvingData::CounterpartyOfferedHTLCOutput(..) => { 1 },
+			PackageSolvingData::CounterpartyReceivedHTLCOutput(..) => { 1 },
+			PackageSolvingData::HolderHTLCOutput(..) => { 2 },
+			PackageSolvingData::HolderFundingOutput(..) => { 3 },
+		};
+		category
```

Comment:
What is category suppose to represent? Is it defined anywhere? Could you use an enum instead?

Reply:
It is supposed to represent an output type. As `type` is a keyword` in Rust I did opt-in for "category". Switched to an enum.